### PR TITLE
REF: move shift logic from BlockManager to DataFrame

### PR DESCRIPTION
**Pull request by jbrockmendel**
Partially-fixes one of the currently-disabled ArrayManager tests.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -5080,28 +5080,45 @@ def shift(
...
-                result = self.iloc[:, -periods:]
-                for col in range(min(ncols, abs(periods))):
-                    # Define filler inside loop so we get a copy
-                    filler = self.iloc[:, -1].shift(len(self))
-                    result.insert(
-                        len(result.columns), label, filler, allow_duplicates=True
-                    )
+                tail = self.iloc[:, -periods:]
+            # pin a simple Index to avoid costly casting
+            tail.columns = range(len(tail.columns))
```
Comment (jreback) [inline pandas/core/frame.py:5100]: are we guaranteed to copy here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/frame.py:5100]: copy what?

Comment (jbrockmendel) [reply pandas/core/frame.py:5100]: oh.  `tail` is not necessarily a copy of the data in `self`, but we do make a copy in the `concat` call below, so the `result` we return is always a copy

Comment (jbrockmendel) [reply pandas/core/frame.py:5100]: `tail` is never the same _object_ as `self`, just may be a view on it

Comment (jreback) [reply pandas/core/frame.py:5100]: got it ok

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
I don't know if it's a particularly useful benchmark, but there is a big slowdown potentially related to this PR: https://pandas.pydata.org/speed/pandas/#frame_methods.Shift.time_shift?python=3.8&Cython=0.29.21&p-axis=1&commits=05a0e98a-bc62e768

**Comment by jbrockmendel** [comment]
best guess is its bc a consolidate is being done by concat, which is called 2x.  easy-ish workaround will be available following #38939

**Comment by jorisvandenbossche** [comment]
I opened an issue to track my above comment -> https://github.com/pandas-dev/pandas/issues/42246

**Pull request merged**